### PR TITLE
feat(sql): add gauge metrics for current cached query count

### LINKED ISSUE #2286: Flaky PGFlushQueryCacheTest.testFlushUpdateCache
### Describe the bug

Failed on linux-arm64:
```
Failed tests: 
  PGFlushQueryCacheTest.testFlushUpdateCache:95->AbstractCairoTest.assertMemoryLeak:384->AbstractCairoTest.assertMemoryLeak:389->AbstractCairoTest.lambda$assertMemoryLeak$1:392->lambda$testFlushUpdateCache$1:128->checkQueryCacheFlushed:142->lambda$checkQueryCacheFlushed$2:144 flush_query_cache() should release native memory: 181018072, 198877656, 283994968
```

### To reproduce

_No response_

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**:
- **OS**:
- **Browser**:
```

### Additional context

_No response_

### LINKED ISSUE #2349: Flaky HttpFlushQueryCacheTest.testJsonQueryFlushQueryCache test
### Describe the bug

Occurred on linux-arm64:
```
2022-07-26T06:38:09.421474Z E i.q.TestListener ***** Test Failed *****io.questdb.cutlass.http.HttpFlushQueryCacheTest.testJsonQueryFlushQueryCache : 
java.lang.AssertionError: flush_query_cache() should release native memory: 45253842, 63113426, 78884370
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at io.questdb.cutlass.http.HttpFlushQueryCacheTest.lambda$testJsonQueryFlushQueryCache$0(HttpFlushQueryCacheTest.java:111)
	at io.questdb.test.tools.TestUtils.assertEventually(TestUtils.java:397)
	at io.questdb.test.tools.TestUtils.assertEventually(TestUtils.java:387)
	at io.questdb.cutlass.http.HttpFlushQueryCacheTest.lambda$testJsonQueryFlushQueryCache$1(HttpFlushQueryCacheTest.java:109)
	at io.questdb.cutlass.http.HttpQueryTestBuilder.lambda$run$0(HttpQueryTestBuilder.java:218)
	at io.questdb.test.tools.TestUtils.assertMemoryLeak(TestUtils.java:462)
	at io.questdb.cutlass.http.HttpQueryTestBuilder.run(HttpQueryTestBuilder.java:77)
	at io.questdb.cutlass.http.HttpFlushQueryCacheTest.testJsonQuery(HttpFlushQueryCacheTest.java:131)
	at io.questdb.cutlass.http.HttpFlushQueryCacheTest.testJsonQueryFlushQueryCache(HttpFlushQueryCacheTest.java:63)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.lang.Thread.run(Thread.java:829)
```

### To reproduce

_No response_

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**:
- **OS**:
- **Browser**:
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Fixes #2286
Fixes #2349

Fixes `flush_query_cache()` function tests flakiness. The tests no longer rely on memory consumption, but instead check the new metric values.

### CODE REVIEW

**Comment by puzpuzpuz** [comment]
@amyshwang this PR adds a number of new Prometheus metrics, so probably it should be on your radar.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 46 / 51 (90.20%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/metrics/MemoryTagGauge.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/metrics/VirtualGauge.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/metrics/NullGauge.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetrics.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/AssociativeCache.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/Metrics.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireMetrics.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpServer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryMetrics.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGJobContext.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/metrics/GaugeImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/HealthCheckMetrics.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/QueryCache.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 50 / 55 (90.91%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/metrics/MemoryTagGauge.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/metrics/VirtualGauge.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/metrics/NullGauge.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetrics.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/AssociativeCache.java|19|19|100.00%|
|:large_blue_circle:|io/questdb/Metrics.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireMetrics.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpServer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryMetrics.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGJobContext.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/metrics/GaugeImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/HealthCheckMetrics.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/QueryCache.java|4|4|100.00%|

**Review by marregui — DISMISSED** [review]
LGTM

**Comment by puzpuzpuz** [comment]
@marregui thanks for the review!

File: core/src/main/java/io/questdb/std/AssociativeCache.java [hunk 1]
```diff
@@ -93,15 +108,26 @@ public CharSequence put(CharSequence key, V value) {
 
         if (Chars.equalsNc(key, keys[lo])) {
             if (values[lo] != value) {
-                Misc.free(values[lo]);
+                if (values[lo] == null) {
+                    cachedGauge.inc();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/std/AssociativeCache.java:112]: could we add comments here please, this is hard to follow logic

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/std/AssociativeCache.java (hunk 1 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/std/AssociativeCache.java:117]: Piece of cake. Added clarifying comments in 83435c8c6393c4086484157fabc24b9920fb36e1

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 51 / 56 (91.07%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/metrics/MemoryTagGauge.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/metrics/VirtualGauge.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/metrics/NullGauge.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetrics.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/AssociativeCache.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/Metrics.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireMetrics.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpServer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryMetrics.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGJobContext.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/metrics/GaugeImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/HealthCheckMetrics.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/QueryCache.java|4|4|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 51 / 56 (91.07%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/metrics/MemoryTagGauge.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/metrics/VirtualGauge.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/metrics/NullGauge.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetrics.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/std/AssociativeCache.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/Metrics.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireMetrics.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpServer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryMetrics.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGJobContext.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/metrics/GaugeImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/HealthCheckMetrics.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/QueryCache.java|4|4|100.00%|
